client/informant/src/lib.rs

```diff
@@ -19,90 +19,138 @@
+use log::{info, trace, warn};
+use parity_util_mem::MallocSizeOf;
+use sc_client_api::{BlockchainEvents, UsageProvider};
+use sc_network::{network_state::NetworkState, NetworkStatus};
+use sp_blockchain::HeaderMetadata;
+use sp_runtime::traits::{Block as BlockT, Header};
+use sp_transaction_pool::TransactionPool;
+use sp_utils::mpsc::TracingUnboundedReceiver;
+use std::fmt::Display;
+use std::sync::Arc;
 use std::time::Duration;
 
 mod display;
 
 /// The format to print telemetry output in.
-#[derive(PartialEq)]
-pub enum OutputFormat {
-	Coloured,
-	Plain,
+#[derive(Clone)]
+pub struct OutputFormat {
+	/// Use colors to display the logs
+	pub colors: bool,
+	/// Add a prefix before every log line
+	pub prefix: String,
 }
 
-/// Creates an informant in the form of a `Future` that must be polled regularly.
-pub fn build(service: &impl AbstractService, format: OutputFormat) -> impl futures::Future<Output = ()> {
-	let client = service.client();
-	let pool = service.transaction_pool();
-
-	let mut display = display::InformantDisplay::new(format);
-
-	let display_notifications = service
-		.network_status(Duration::from_millis(5000))
-		.for_each(move |(net_status, _)| {
-			let info = client.usage_info();
-			if let Some(ref usage) = info.usage {
-				trace!(target: "usage", "Usage statistics: {}", usage);
-			} else {
+macro_rules! build {
+	($client:expr, $network_status_stream_builder:expr, $pool:expr, $format:expr) => {{
+		let mut display = display::InformantDisplay::new($format.clone());
+
+		let client_1 = $client.clone();
+		let display_notifications = $network_status_stream_builder(Duration::from_millis(5000))
+			.for_each(move |(net_status, _)| {
+				let info = client_1.usage_info();
+				if let Some(ref usage) = info.usage {
+					trace!(target: "usage", "Usage statistics: {}", usage);
+				} else {
+					trace!(
+						target: "usage",
+						"Usage statistics not displayed as backend does not provide it",
+					)
+				}
+				#[cfg(not(target_os = "unknown"))]
 				trace!(
 					target: "usage",
-					"Usage statistics not displayed as backend does not provide it",
-				)
+					"Subsystems memory [txpool: {} kB]",
+					parity_util_mem::malloc_size(&*$pool) / 1024,
+				);
+				display.display(&info, net_status);
+				future::ready(())
+			});
+
+		let mut last_best = {
+			let info = $client.usage_info();
+			Some((info.chain.best_number, info.chain.best_hash))
+		};
+
+		let display_block_import = $client.import_notification_stream().for_each(move |n| {
+			// detect and log reorganizations.
+			if let Some((ref last_num, ref last_hash)) = last_best {
+				if n.header.parent_hash() != last_hash && n.is_new_best  {
+					let maybe_ancestor = sp_blockchain::lowest_common_ancestor(
+						&*$client,
+						last_hash.clone(),
+						n.hash,
+					);
+
+					match maybe_ancestor {
+						Ok(ref ancestor) if ancestor.hash != *last_hash => info!(
+							"♻️  Reorg on #{},{} to #{},{}, common ancestor #{},{}",
```

Comment:
Prefix? And also does not checks if color output is enabled.

Reply:
Prefix => ok

> And also does not checks if color output is enabled.

That I wanted to discuss with you. I propose that we drop the color check entirely and use only colors. The reason is that the colors are strip anyway. It will reduce the code size and maintenance a bit.